app/components/market/submit/submit-market-application.component.ts

```diff
@@ -75,134 +17,161 @@ export class SubmitMarketApplicationComponent implements OnInit {
+            this.templateService.getTemplates(this.applicationId, this.actionId).subscribe((templates) => {
+                this.jobTemplate = templates.job;
+                this.poolTemplate = templates.pool;
+                this._parseParameters();
+                this._createForms();
+            });
+            this.templateService.getApplication(this.applicationId).subscribe((application) => {
+                this.icon = application.icon;
             });
         });
     }
 
-    public getContainerFromFileGroup(fileGroup: string) {
-        return fileGroup && `fgrp-${fileGroup}`;
+    public pickMode(mode: Modes) {
+        this.modeState = mode;
     }
 
     @autobind()
     public submit() {
         this.error = null;
-        const jobParams = this.jobFormGroup.value;
-        // RPC takes in Template JSON object and Parameter JSON object
-        const obs = this.pythonRpcService.callWithAuth("submit-ncj-job", [this._buildJobTemplate(), jobParams]);
-        obs.subscribe({
-            next: (data) => this._redirectToJob(data),
-            error: (err) => this.error = ServerError.fromPython(err),
-        });
+        let obs;
+        switch (this.modeState) {
+            case Modes.NewPoolAndJob: {
+                obs = this.pythonRpcService.callWithAuth("expand-ncj-pool", [this.poolTemplate, this.poolParams.value])
+                    .cascade((data) => this._runJobWithPool(data));
+                break;
+            }
+            case Modes.OldPoolAndJob: {
+                this.jobTemplate.job.properties.poolInfo = this.pickedPool.value;
+                obs = this.pythonRpcService.callWithAuth("submit-ncj-job", [this.jobTemplate, this.jobParams.value])
+                    .cascade((data) => this._redirectToJob(data.properties.id));
+                break;
+            }
+            case Modes.NewPool: {
+                obs = this.pythonRpcService.callWithAuth("create-ncj-pool", [this.poolTemplate, this.poolParams.value])
+                    .cascade((data) => this._redirectToPool(data.id));
+                break;
+            }
+            default: {
+                return obs;
+            }
+        }
+        if (obs) {
+            obs.subscribe({
+                error: (err) => this.error = ServerError.fromPython(err),
+            });
+        }
         return obs;
     }
 
-    private _redirectToJob(data) {
-        const jobId = data.properties.id;
-        this.router.navigate(["/jobs", jobId]);
+    private _parseParameters() {
+        const jobParameters = this.jobTemplate.parameters;
+        const jobTempWrapper: any[] = [];
+        for (let name of Object.keys(jobParameters)) {
+            const param = jobParameters[name];
+            jobTempWrapper.push(new NcjParameterWrapper(name, param));
+        }
+        this.jobParametersWrapper = jobTempWrapper;
+        const poolParameters = this.poolTemplate.parameters;
+        const poolTempWrapper: any[] = [];
```

Comment:
shouldn't this be 'let' instead of 'const' if you are going to append to it later?

Reply:
https://medium.com/javascript-scene/javascript-es6-var-let-or-const-ba58b8dcde75

![image](https://user-images.githubusercontent.com/30036553/29788235-b5384760-8be6-11e7-89de-cda5eb20398b.png)

Since the array variable is not being reassigned then it's safe to use const. Even though its contents are changing.